tests/cxx/badinc.cc

```diff
@@ -1119,18 +1119,16 @@ int main() {
   // IWYU: I1_PtrDereferenceClass is...*badinc-i1.h
   local_i1_ptrdereference_class->a();
 
-  // Calling an overloaded function.  In the first two cases,
```

Comment:
This looks like a regression to me; would it be warranted to add a TODO here to explain that this is not ideal, and what's blocking us from improving?

Or am I misunderstanding, and this is actually better?

Reply:
The primary reason of the changes was to fix reporting ADL-found overloads at the use side. The problem is because when all the non-ADL overloads are in the same file, the calling expression appears in `processed_overload_locs` and is ignored. Fixing the bug while keeping the current logic would introduce additional complexity, but then I realized that the current logic cannot work well in all cases: in general, while scanning a user of `badinc.h` IWYU cannot in principle say that it will actually suggest to include `badinc-i1.h` in `badinc.h` while fixing `badinc.h`, as far as it doesn't analyze all the translation units simultaneously (see the example in #1525).

Reply:
Thanks for the summary. I suggest adding something like:
```
  // TODO: CallOverloadedFunction should really be responsible for its
  // implementation details, in this case which function overload is called. We
  // don't yet handle the combination of overload resolution and template
  // instantiation with full fidelity.
```
just to make it clear that the tested behavior is not really the desired behavior.

Reply:
My point is that it's not achievable for IWYU. IWYU _should_ suggest `#include` in such cases in general, because it doesn't know whether it will actually suggest it in tpl-defining header or not.

Reply:
I see. So the old behavior was a heuristic that masked that fact in some cases, but also caused other issues... Right?

I know better than to argue with you about template analysis in IWYU :slightly_smiling_face:, but these attributions look obviously wrong to me, and I'm concerned that they will make IWYU unusable. So the improvements come for instantiation-side ADL resolution -- I don't have any intuition for how much more common that is than non-ADL lookup on definition side. It's essentially these two constructs played out against each other, right?

To cover both cases, it seems like we need some kind of author-intent hint or pragma, then, eventually?

Reply:
I had some time away from the keyboard, and just wanted to clarify: it's not super important to me what the eventual behavior is, but my priority is to make sure there's enough information available in the tests and code that someone sitting down with this problem later will have some context for why IWYU behaves this way.

So I don't want to block this patch on functional concerns, I just want to make sure there's some rationale for _why_ `CallOverloadedFunction.*` don't claim ownership of the implementation-side functions.

Reply:
> So the old behavior was a heuristic that masked that fact in some cases, but also caused other issues... Right?

It was a heuristic, but rather the issues were just not fully understood.

> I'm concerned that they will make IWYU unusable

That would be a strong statement! IWYU with the similar patch was successfully used at my previous work.

> I don't have any intuition for how much more common that is than non-ADL lookup on definition side

Anyway, IWYU should produce compilable code, hence some redundancy is better (assuming the presence of double-inclusion guards in headers, of course).

> To cover both cases, it seems like we need some kind of author-intent hint or pragma, then, eventually?

It works already: just include the function-defining header in the template-defining header directly so that IWYU doesn't suggest `#include` at the instantiation side. I even stated that in [my original commit](https://github.com/bolshakov-a/include-what-you-use/commit/45ae521e301aaf29c073d32a2803bb24fa671ec4#diff-b4ba3b28c8dd51625839eed7fa55c371893205bb3ed9142b7884a913dc68d129R1141). Probably, the comments should be copied from there?

Reply:
> Anyway, IWYU should produce compilable code, hence some redundancy is better (assuming the presence of double-inclusion guards in headers, of course).

This is a very fair point! 

> It works already: just include the function-defining header in the template-defining header directly so that IWYU
> doesn't suggest #include at the instantiation side. I even stated that in
> [my original commit](https://github.com/bolshakov-a/include-what-you-use/commit/45ae521e301aaf29c073d32a2803bb24fa671ec4#diff-b4ba3b28c8dd51625839eed7fa55c371893205bb3ed9142b7884a913dc68d129R1141). Probably, the comments should be copied from there?

Yeah, that would make sense to me -- those comments help clarify a lot.